cache/cache.go

```diff
@@ -187,21 +187,63 @@ func (c *Cache) Set(key uint64, value interface{}, cost int64) {
 	}
 }
 
+// GetOrCompute returns the value of key. If there is no such key, it will compute the
+// value using the factory function `f`. If there are concurrent call on same key,
+// the factory function will be called only once.
+func (c *Cache) GetOrCompute(key uint64, f func() (interface{}, int64, error)) (interface{}, error) {
+	if c == nil {
+		return nil, nil
+	}
+	for {
+		i := c.store.GetOrNew(key)
+		if v := i.value.Load(); v != nil {
+			return v, nil
+		}
+
+		i.Lock()
+		if i.dead {
+			i.Unlock()
```

Comment:
Can we extract a method and use `defer`, since defer cost is trivial on the latest Go version.